File: javascript/packages/linkseal-resolver/src/__tests__/PolicyProcessor.test.ts
import { describe, it, expect } from 'vitest';
import { PolicyProcessor } from '../PolicyProcessor.js';
import type { PolicyConfig } from 'linkseal';

describe('PolicyProcessor', () => {
  describe('process', () => {
    it('passes through resolved variables', () => {
      const result = new PolicyProcessor().process({ token: 'abc' }, { missing: 'error' });
      expect(result.result).toEqual({ token: 'abc' });
      expect(result.missing).toEqual([]);
    });

    // Regression: missing was computed from resolved keys only, so it was always empty.
    it('reports null variables as missing', () => {
      const result = new PolicyProcessor().process(
        { userId: 'user_123', token: null },
        { missing: 'error' }
      );
      expect(result.missing).toEqual(['token']);
      expect(result.result).toEqual({ userId: 'user_123' });
    });

    it('fills null variables from defaults in default mode', () => {
      const result = new PolicyProcessor().process(
        { locale: null },
        { missing: 'default', defaults: { locale: 'zh-CN' } }
      );
      expect(result.result).toEqual({ locale: 'zh-CN' });
      expect(result.missing).toEqual([]);
    });

    it('still reports missing when default mode lacks a matching default', () => {
      const result = new PolicyProcessor().process(
        { locale: null, token: null },
        { missing: 'default', defaults: { locale: 'zh-CN' } }
      );
      expect(result.result).toEqual({ locale: 'zh-CN' });
      expect(result.missing).toEqual(['token']);
    });

    it('resolved variables take precedence over defaults', () => {
      const result = new PolicyProcessor().process(
        { locale: 'en-US' },
        { missing: 'default', defaults: { locale: 'zh-CN' } }
      );
      expect(result.result).toEqual({ locale: 'en-US' });
    });

    it('ignores defaults outside of default mode', () => {
      const result = new PolicyProcessor().process(
        { locale: null },
        { missing: 'ignore', defaults: { locale: 'zh-CN' } }
      );
      expect(result.result).toEqual({});
      expect(result.missing).toEqual(['locale']);
    });

    it('does not mutate the supplied config', () => {
      const config: PolicyConfig = { missing: 'error' };
      new PolicyProcessor().process({ token: null }, config);
      expect(config).toEqual({ missing: 'error' });
      expect(config.defaults).toBeUndefined();
    });
  });

  describe('applyMissingPolicy', () => {
    it('error mode reports missing variables', () => {
      const result = new PolicyProcessor().applyMissingPolicy(
        { userId: 'user_123', token: null },
        { missing: 'error' }
      );
      expect(result.error?.code).toBe('MISSING_VARIABLE');
      expect(result.error?.message).toContain('token');
    });

    it('error mode succeeds when all variables resolve', () => {
      const result = new PolicyProcessor().applyMissingPolicy(
        { token: 'abc', locale: 'en' },
        { missing: 'error' }
      );
      expect(result.error).toBeUndefined();
      expect(result.result).toEqual({ token: 'abc', locale: 'en' });
    });

    it('ignore mode omits missing variables without erroring', () => {
      const result = new PolicyProcessor().applyMissingPolicy(
        { userId: 'user_123', token: null },
        { missing: 'ignore' }
      );
      expect(result.error).toBeUndefined();
      expect(result.result).toEqual({ userId: 'user_123' });
    });

    it('default mode uses provided defaults', () => {
      const result = new PolicyProcessor().applyMissingPolicy(
        { theme: 'light', lang: null },
        { missing: 'default', defaults: { theme: 'dark', lang: 'en' } }
      );
      expect(result.error).toBeUndefined();
      expect(result.result).toEqual({ theme: 'light', lang: 'en' });
    });

    it('empty variable set is not an error', () => {
      const result = new PolicyProcessor().applyMissingPolicy({}, { missing: 'error' });
      expect(result.error).toBeUndefined();
      expect(result.result).toEqual({});
    });
  });
});

File: java/src/test/java/io/github/xesam/linkseal/PolicyProcessorTest.java
package io.github.xesam.linkseal;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.github.xesam.linkseal.model.MissingPolicy;
import io.github.xesam.linkseal.model.PolicyConfig;
import io.github.xesam.linkseal.resolver.PolicyProcessor;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

final class PolicyProcessorTest {

  private final PolicyProcessor processor = new PolicyProcessor();

  private static Map<String, String> vars(String... keyValues) {
    Map<String, String> map = new LinkedHashMap<>();
    for (int i = 0; i < keyValues.length; i += 2) {
      map.put(keyValues[i], keyValues[i + 1]);
    }
    return map;
  }

  @Test
  void passesThroughResolvedVariables() {
    var result = processor.process(vars("token", "abc"), new PolicyConfig(MissingPolicy.ERROR));
    assertEquals(Map.of("token", "abc"), result.result());
    assertEquals(List.of(), result.missing());
  }

  /** Regression: missing was derived from resolved keys only, so it was always empty. */
  @Test
  void reportsNullVariablesAsMissing() {
    var result = processor.process(
        vars("userId", "user_123", "token", null),
        new PolicyConfig(MissingPolicy.ERROR)
    );
    assertEquals(List.of("token"), result.missing());
    assertEquals(Map.of("userId", "user_123"), result.result());
  }

  @Test
  void defaultModeFillsNullVariables() {
    var result = processor.process(
        vars("locale", null),
        new PolicyConfig(MissingPolicy.DEFAULT, Map.of("locale", "zh-CN"))
    );
    assertEquals(Map.of("locale", "zh-CN"), result.result());
    assertEquals(List.of(), result.missing());
  }

  @Test
  void defaultModeStillReportsMissingWithoutMatchingDefault() {
    var result = processor.process(
        vars("locale", null, "token", null),
        new PolicyConfig(MissingPolicy.DEFAULT, Map.of("locale", "zh-CN"))
    );
    assertEquals(Map.of("locale", "zh-CN"), result.result());
    assertEquals(List.of("token"), result.missing());
  }

  @Test
  void resolvedVariablesTakePrecedenceOverDefaults() {
    var result = processor.process(
        vars("locale", "en-US"),
        new PolicyConfig(MissingPolicy.DEFAULT, Map.of("locale", "zh-CN"))
    );
    assertEquals(Map.of("locale", "en-US"), result.result());
  }

  @Test
  void ignoreModeDoesNotApplyDefaults() {
    var result = processor.process(
        vars("locale", null),
        new PolicyConfig(MissingPolicy.IGNORE, Map.of("locale", "zh-CN"))
    );
    assertEquals(Map.of(), result.result());
    assertEquals(List.of("locale"), result.missing());
  }

  @Test
  void errorModeThrowsWhenMissing() {
    var error = assertThrows(
        IllegalArgumentException.class,
        () -> processor.applyMissingPolicy(vars("token", null), new PolicyConfig(MissingPolicy.ERROR))
    );
    assertTrue(error.getMessage().contains("token"));
  }

  @Test
  void errorModeSucceedsWhenAllPresent() {
    assertDoesNotThrow(() -> processor.applyMissingPolicy(
        vars("token", "abc", "locale", "en"),
        new PolicyConfig(MissingPolicy.ERROR)
    ));
  }

  @Test
  void ignoreModeOmitsMissingWithoutThrowing() {
    Map<String, String> result = processor.applyMissingPolicy(
        vars("userId", "user_123", "token", null),
        new PolicyConfig(MissingPolicy.IGNORE)
    );
    assertEquals(Map.of("userId", "user_123"), result);
  }

  @Test
  void defaultModeUsesProvidedDefaults() {
    Map<String, String> result = processor.applyMissingPolicy(
        vars("theme", "light", "lang", null),
        new PolicyConfig(MissingPolicy.DEFAULT, Map.of("theme", "dark", "lang", "en"))
    );
    assertEquals(Map.of("theme", "light", "lang", "en"), result);
  }

  @Test
  void emptyVariableSetIsNotAnError() {
    Map<String, String> result = processor.applyMissingPolicy(Map.of(), new PolicyConfig(MissingPolicy.ERROR));
    assertEquals(Map.of(), result);
  }
}
